< 과제 3>   2018136085 이세희
===

1번
=== 
L=[1, 2, 3, 4, 5]일 때 다음 각 문장을 수행한 후의 결과를 보고 납득할 만한 이유를 설명하시오.

(1)
---

In [2]:
L=[1, 2, 3, 4, 5]
L[1:3] = [100]
print(L)

[1, 100, 4, 5]


<설명>

슬라이싱으로 L[1:3]은 2,3 항목 이다.  따라서 2,3대신 100 항목을 넣는다는 이야기이다. 그리고 여기선, 항목간 복사를 진행하므로 [2,3] => [100]으로 바뀌어 [int(100)]인 항목으로 들어가게 된 것이다. 

따라서, 1과 100, 4 , 5 만 남은 것이다. 

(2)
---

In [38]:
L=[1, 2, 3, 4, 5]
L[1:3] = [(100)]
print(L)

[1, 100, 4, 5]


< 설명 >

흔히 ()는 튜플로 많이 생각하는데, 
여기서는 ()에 의미는 그저 우선순위 수학 계산식의 ( ) 기능과 같다. 따라서 리스트에 100으로 저장 된다.
그리고 슬라이싱은 항목간 복사를 진행한다. 따라서 [ ] 안에 있는 100 항목이 입력된다. 


(3)
==

In [2]:
L=[1, 2, 3, 4, 5]
L[1:3] = 100
print(L)

TypeError: can only assign an iterable

<설명>
리스트 슬라이싱은 iterable 한 값에 일부분을 출력하는 기능으로써, 따라서 iterable 하지 않으면 대입할 수 없다. 

따라서 대입하고 싶으면 겉 부분에 list []를 씌워서 iterable하게 만들어야 한다. 

100은 그저 int형 100 이므로 iterable 하지 않아서 오류가 발생했다.  

(4)
--

In [36]:
L=[1, 2, 3, 4, 5]
L[1:3] = (100)
print(L)

TypeError: can only assign an iterable

<설명> 
이도 마찬가지로 여기서 (100)은 int 100과 같은 값으로 위와 동일하게 오류가 발생한다.  여기서의 괄호는 계산 우선 순위 괄호를 의미하므로 
(100)은 100과 같다.

(5)
---

In [16]:
L=[1, 2, 3, 4, 5]
L[1:3] = (100, 101, 102)
print(L)

[1, 100, 101, 102, 4, 5]


<설명>

슬라이싱은 항목간 복사가 된다. 근데, 여기서 항목 하나씩 값을 복사하게되면, L[1] = 100 을 넣기 , L[2] = 101 넣기  L하나 더 넣고 밀기 해서 한개 씩 바라보게 되어서 int 값으로 대입했기 때문이라고 생각한다.

그래서 만약 
L[1] =  (100, 101, 102) 을 하면 튜플로 저장된다. 

(6)
---

In [17]:
L=[1, 2, 3, 4, 5]
L[1:3] = [100, 101, 102]
print(L)

[1, 100, 101, 102, 4, 5]


<설명> 

여기서도 그냥 리스트로 넣기 때문에 항목간 복사를 실시한다. 따라서 [1,[100, 101 , 102],4, 5 ]  아닌 다음과 같은 결과 값이 나왔다. 

(7)
---

In [20]:
L=[1, 2, 3, 4, 5]
L[1:3] = [(100, 101, 102)]
print(L)

[1, (100, 101, 102), 4, 5]


<설명> 

여기서 대입 연산자 오른쪽은 [(100, 101, 102)] 이다 이는 A=   [(100, 101, 102)] 일 때, A[0] = (100, 101, 102)이다,

슬라이싱은 항목간 복사이다 그래서 A[0]의 값을 한번에 대입하므로 튜플로 들어가게 된다. 


(8)
---

In [21]:
L=[1, 2, 3, 4, 5]
L[1:3] = [[100, 101, 102]]
print(L)

[1, [100, 101, 102], 4, 5]


<설명>

이 문제도 마찬가지다 [] 안에 인덱싱이 [0]으로 되므로 항목간 복사를 실시한다. 그래서 [0] = [100, 101, 102] 뭉텅이가 한번에 입력되는 것이다. 

2번
===
* 문자열 S = 'Hello World and Python'에 대해 다음 요구사항에 대한 Python 코드를 제시하시오.
* 1) 단어의 순서가 역순으로 되어진 문자열 ('Python and World Hello')을 만들고
* 2) 1)의 결과 문자열에 대해 메소드 split과 join을 이용하여 공백을 모두 없엔 문자열 'PythonandWorldHello'을 만드시오.

In [88]:
S = 'Hello World and Python'  #원본 


#1) 

S = S.split()    #거꾸로 넣기 위해 분해 작업 부터시작 ( 문자열 -> 리스트로 변경)
S.reverse()      #문자의 요소 순서를 바꿈 거꾸로 

S=' '.join(a for a in S)   #띄어쓰기 한 칸씩 널고 다시 문자열로 변경 

print(S)





#2)
S = S.split() #다시 분해해서 띄어쓰기로 값을 나눈 후  
S="".join(S)  #여기서 나온 값들을 공백을 없앤 후 출력 
print(S)

Python and World Hello
PythonandWorldHello


<설명> 

(1)

주석에서도 거의 설명을 했지만, 먼저 문제를 보면 완전히 글자 하나하나 거꾸로 출력하는 것이 아니다. 
따라서 인덱싱에 위치를 0 1 2 3 4를 reverse하여 4 3 2 1 0 할 수 있는 것을 이용하기 위해 단어를 하나의 요소를 만들기로 생각했다.
그래서, Split을 통해서 단어 마다 요소로 나누었고, 이 값을 reverse 했고, 분리된 값들을 하나로 모으기 위해서 join을 사용해 완성했다.

※여기서, 주의할 점은  Split()은 리스트로 바뀌고 join은 문자열로 바뀜을 주의해야 한다. 


(2)

2번째도, 같은 방식으로 Split과 join을 사용하라고 했으니, 다시 분해해서 다시 join하는 작업을 걸쳐서 출력했다. 

3번
===
> 다음 코드를 보고 물음에 답하시오.

> a = [1, 2, 3]

> b = a * 3

> c = [a] * 3

> print(b)

> print(c)


(1)
---
* 위 코드를 수행한 이후 다음 코드를 수행한 후 b와 c의 내용을 확인하시오.
* a[0]=0

In [22]:
a = [1, 2, 3]
b = a * 3
c = [a] * 3

a[0] = 0

print(b)
print(c)

[1, 2, 3, 1, 2, 3, 1, 2, 3]
[[0, 2, 3], [0, 2, 3], [0, 2, 3]]


<설명>

여기서 우리가 기대 하는 것은 분명 a값은 이전에 대입 했는데, a[0] = 0을 하고 나니, 값이 C에 값이 바뀌었다. 이는 왜 그런 것인가 ?

 _* 연산을 할 때 항목 하나씩 하게 되어서 b에는 int형으로 그저 값이 복사되어 입력된다.

하지만, C에서는 []로 묶여있어서 래퍼런스를 복사하게 된다. 따라서 a의 래퍼런스를 c[0]에 하나 c[1],c[2]까지 넣어주게 된다. 

래퍼런스값이기 때문에 뒤에서 a[0] =0 값을 해주면 역시 c 또한 값이 바뀌게 된다.

(2)
----
b와 c 객체의 차이는 무엇인가?

In [33]:

a = [1, 2, 3]
b = a * 3
c = [a] * 3

a[0] = 0


print(id(a[0]))  #a와 b 주소 비교 
print(id(b[0]))
 
print(id(a))    #a와 c 주소 비교 
print(id(c[0]))


140729259534096
140729259534128
2017357854656
2017357854656


<설명> 

여기서, ID는 가르키는 주소를 반환하는 함수이다. 이를 이용해서 확인해보자 

먼저, b와 a를 비교해보자, 같은 값인 1을 b[0]과 a[0]을 비교해보자, 

>140729259534096

>140729259534128

다른 값을 출력하는 것을 볼 수 있다. 따라서 입력 받을 때 래퍼런스가 아닌 값 복사를 통해서 전달했기 때문에 나중에 원본 값을 바꿔도
아무런 변화가 없다. 

하지만, 여기서 a와 c[0]을 보면 같은 주소를 가르킨다. 따라서 c = [a]를 넣는 과정에서 값만 복사하는 것이 아니라 래퍼런스를 복사하게 된다.

>2017357854656

>2017357854656

따라서, 같은 주소를 참조하고 있기 때문에 하나라도 값이 바뀌면 a,c 모두 값이 변경되게 된다. 

4번
===
* 다음 문자열을 ':'을 기준으로 분리하여 리스트로 만들고 각 문자열의 좌우 공백을 제거하시오 (즉, 문자열 S에서 l을 만들어라)
* s = '  first star   :   second star   :    third star  '
* l = ['first star', 'second star', 'third star']
* 1) for ~ in 문을 이용한 코드를 작성하시오.
* 2) 리스트 내포(list comprehension)를 이용한 코드를 작성하시오.

In [70]:
s = ' first star : second star : third star '
s= s.split(':')  #여기서 문자열 -> 리스트로 만들어짐 



l=[] #값을 저장할 리스트를 만들고 
l = [i.strip() for i in s]  #리스트 내포 사용

print(l)

['first star', 'second star', 'third star']


<설명> 

여기서, split으로 : 을 삭제 했고, 그 다음 새로운 리스트에 원래는 

for i in s:
    i = i.strip()
    l.append(i)
    
의 과정이 필요하다. 하지만, 여기서는 리스트 내포를 이용해서 간결하게 한줄로 코딩할 수 있고, 
리스트 내포를 통해서 append 도 사용할 필요없이 간결하게 구성했다. 

5번
===
* 다음과 같이 0보다 큰 정수 리스트 변수 list를 인자로 받는 함수 addall(list)와 addallodd(list)를 for ~ in 리터널과 리스트 내포 방식으로 각각 제시하시오.
* 1) addall(list) 함수 (리스트 내의 모든 정수를 더하는 함수로서 해답으로는 for ~ in 리터럴과 리스트 내포 방식으로 각각 제시하시오.)
* 2) addallodd(list) 함수 (리스트내의 모든 홀수를 더하는 함수로서 해답으로는 for ~ in 리터럴과 리스트 내포 방식으로 각각 제시하시오.)


In [66]:
def addall(L):
    a = 0
    
    
#for in 을 사용한 방법 
    #for i in L:
    #    a+=i
    

#내포 방식 

    a = [i+a for i in L]
    return a
    
addall([1,2,3,4,5,6,7,8,9])    

[1, 2, 3, 4, 5, 6, 7, 8, 9]

6번
===
* 다음 코드를 보고 물음에 답하시오.



In [67]:
L1 = [1, 2, 3]
L2 = [4, 5, 6]
d = {'low':L1, 'high':L2}
e = d
f = d.copy()

print(d)
print(e)
print(f)
print()
d['low'] = [10, 20, 30]
d['high'][1] = 500
print(d)
print(e)
print(f)

{'low': [1, 2, 3], 'high': [4, 5, 6]}
{'low': [1, 2, 3], 'high': [4, 5, 6]}
{'low': [1, 2, 3], 'high': [4, 5, 6]}

{'low': [10, 20, 30], 'high': [4, 500, 6]}
{'low': [10, 20, 30], 'high': [4, 500, 6]}
{'low': [1, 2, 3], 'high': [4, 500, 6]}


* 1) 위 코드의 수행 결과 d와 e는 항상 같은 결과를 출력한다. 그 이유는 무엇인가?
* 2) 위 코드의 수행 결과 마지막 f의 내용은 다음과 같다. 이 때 'high' 키의 값 내용 중 500이 출력된 이유와 'low' 키의 값 내용 중 [1, 2, 3]이 출력된 이유를 설명하시오 ([hint]: shallow copy 관점에서 설명하시오.)


* {'high': [4, 500, 6], 'low': [1, 2, 3]}

<설명> 

(1) 
---
e = d 대입 연산자는  값이 할당되는 것이 아니라 같은 메모리 주소를 바라봅니다. 따라서, 원본 값이 
변경 되면 똑같이 바뀌게 되는 것입니다. (ex,C언어의 포인터) 

(2) 
---
여기서는 얕은 복사가 진행되었습니다. 따라서 새로운 래퍼런스를 만들어 복사 합니다. 여기까지는 아무 문제가 없고, 

따라서 여기서 d['low'] 의 값을 바꾸어도 변경되지 않습니다. 

하지만, 문제는 f안에 있는 내부까지 복사를 하지 못하는 문제가 있습니다. 
따라서, F안에 있는 리스트는 아직 D안에 있는 리스트 각각의 값들을 래퍼런스로 참조하고 있기 때문에 

F 리스트 항목에 접근하면 원본 값도 바뀌게 되는 것 입니다. 

 ※ 완전하게 복사를 진행하고 싶다면 deepcopy를 통해서 객체 내부의 있는 객체까지 복사를 진행해야 합니다.  


7번
===
* 사전 d = {'one':1, 'two':2, 'three':3, 'four':4, 'five':5}가 주어졌을 때 다음 요구사항에 맞는 코드를 제시하시오
* [참고]: d.keys(), d.values()를 통해 리스트를 얻어낸 후 리스트가 지니고 있는 sort(cmp=None, key=None, reverse=False)함수를 활용하시오.


* 1) 키의 알파벳 오름차순 순서대로 튜플 (키, 값)을 차례대로 출력하시오.
* 2) 키의 알파벳 내림차순 순서대로 튜플 (키, 값)을 차례대로 출력하시오.
* 3) 값의 오름차순 순서대로 튜플 (키, 값)을 차례대로 출력하시오.
* 4) 값의 내림차순 순서대로 튜플 (키, 값)을 차례대로 출력하시오.

In [111]:
d_1 = {'one':1, 'two':2, 'three':3, 'four':4, 'five':5}


a_1 = list(zip(d.keys(),d_1.values()))


a_1.sort(reverse=False, key=None)
print("키의 오름차순 정렬: " ,a_1)


a_1.sort(reverse=True, key=None)
print("키의 내림차순 정렬: " ,a_1)


a_1.sort(key=lambda d_1: d_1[1],reverse=False)
print("값의 오름차순 정렬 : " , a_1)


a_1.sort(key=lambda d_1: d_1[1],reverse=True)
print("값의 내림차순 정렬 : " , a_1)

TypeError: 'list' object is not callable

<설명> 

sort()는 리스트에만 사용할 수 있기 때문에 sort를 사용하기 위해서 먼저, 정렬 해 주고 list를 통해서 리스트를 만들어서 sort를 사용했다. 


8번
===
* 다음 문제를 정확히 읽고 적합한 코딩을 제시하시오.
* length(X)를 다음과 같이 정의한다 : 10진수로 표현된 X의 자릿수
* Strange라는 숫자를 아래과 같이 정의한다.
* 음이 아닌 정수 중 자릿수가 1인 모든 숫자는 Strange이다.
* length(X) > 1 인 숫자 X 는 다음 2가지를 만족하면 Strange이다.
* X는 length(X)로 정확히 나누어 떨어진다.
* X / length(X) 가 strnage이다.

* 이 문제에서 숫자 하나가 주어질 때, 이 숫자가 Strange인지 판단하는 함수 isStrange(num)을 작성하시오.
* isStrange(num) 함수가 재귀적인 함수가 되도록 작성해 보세요.

In [114]:
def isStrange(num):

    
    

    if len(str(num))==1:
        return True
    
    elif num % len(str(num)) !=0 :
        return False
    elif num % len(str(num)) == 0:
        return isStrange(num % len(str(num)))
    
    

    
    
    
isStrange(6012302549522098651)


False

<설명> 


여기서, 나누어서 나누어 떨어진다면 재귀함수를 이용해서 나눈 값을 다시 불러 계산한다. 

나누어 떨어지지 않았다면 isStrange가 아니므로 False를 반환한다. 

계속 맞다면 계속 재귀함수를 통해 계속 호출한다.

9번
===
* 다음 문제를 정확히 읽고 적합한 코딩을 제시하시오.

* 1번 그룹, 2번 그룹, ... n번 그룹으로 번호가 매겨진 총 n개의 그룹이 존재하고, ith 그룹에는 임의의 팀원 수 ai 명이 있다.
* 하나의 버스 정류장이 있고, 버스는 오직 한 대만 운영한다. 버스는 버스 정류장에 도착하고 모든 그룹은 버스를 통해 여행하길 원한다.
* 하지만, 같은 그룹에 속한 팀원들은 분리되기를 원치 않는다. 그래서 그룹의 모든 팀원이 탈 수 있을 때 에만 그 그룹은 버스에 탑승한다.
* 또한, 이들은 그룹의 번호순으로 여행을 하길 원한다. 예를 들어, 3번 그룹은 1번 그룹과 2번 그룹이 여행을 한 뒤에 버스를 탈 수 있거나, 1번 * 그룹과 2번 그룹이 이미 버스에 탑승한 상태에 있을 때 동시에 버스를 탈 수 있다.
* 마지막으로, 버스는 빈 좌석이 없을 때 에만 여행을 떠난다.


* 이 문제에서 각 그룹의 팀원 수가 리스트로 주어질 때, 가능한 모든 버스의 사이즈 x(x명의 사람들이 동시에 탈 수 있음)를 반환하는 함수 solve(a) 를 작성하시오.
* [주의] 반드시 계산 복잡도를 최대한 낮추세요. 아래 예시 3번을 실행했을 때 결과가 바로 나타날 수 있도록!!

In [82]:
list = [9999] * 10**4

result = [] 
count =0 
range_min=0
range_max=1



n=sum(list)
divisorsList = []

for i in range(1, int(n**(1/2)) + 1):   #약수를 구하는 부분 
    if (n % i == 0):
        divisorsList.append(i) 
        if ( (i**2) != n) : 
            divisorsList.append(n // i)
            
divisorsList.sort()


range_result = [i for i in divisorsList if i>= max(list)]  #약수 중에서 최소 값 정해주기



while (count < len(range_result) ):    #더하면서 순위 맞는 지 검사하는 부분

        b=list[range_min:range_max] 

        if range_result[count] == sum(b):   #그룹 순위대로 차에 끊기 
            range_min=range_max
            range_max=range_max+1 


            if (range_min == len(list)):     #끝까지 검사했는데 딱 맞아 떨어졌을 때 
                result.append(range_result[count]) 
                
                
                count +=1
                range_min=0
                range_max=1




        else:    #아니면 범위를 늘리자
            
            if (sum(b) > range_result[count]):  #시간 복잡도 줄이기 만약 더하면서 검사하는데, 범위를 초과하면 뒤는 검사 x 하고 다음 
                count +=1
                range_min=0
                range_max=1
                
            range_max+=1
          

        
print(result)


[9999, 19998, 39996, 49995, 79992, 99990, 159984, 199980, 249975, 399960, 499950, 799920, 999900, 1249875, 1999800, 2499750, 3999600, 4999500, 6249375, 9999000, 12498750, 19998000, 24997500, 49995000, 99990000]


<설명>
먼저, 문제부터 확인하겠다. 
문제를 해석하고 나서 찾아낸 조건이 있다.
1. 버스에 태우는 최소값은 max(list) 이다. 
2. 버스에 태우는 최대값은 sum(list) 이다.
3. 앞 순서대로 버스에 태우면서 나머지가 없는 지 확인해야 한다.  
4. 최대값= sum(list)의 약수이다. 

따라서, 해당 조건을 모두 만족해야 한다.

다음 조건을 통해서 코딩을 해보자. 

------

1. 모든 값을 대입해보면서 버스 사이즈X를 구하는 것은 시간 복잡도를 매우 높게 하는 코드이다. 따라서, 4번 조건을 통해서 우리가 예상할 수 있는 버스 사이즈를 구해놓자,  

 >예를 들어 [1, 2, 1, 1, 1, 2, 1, 3] 를 보면, 최대값의 약수는 12의 약수이다. [1,2,3,4,6,12] 


2. 이제 약수들 중에서 최소값과 최대값 범위를 지정해서 더욱 범위를 줄인다. 

>최소 3이상 최대 12이하 조건 만족하는 약수 = [3,4,6,12]

3. 이제 버스에 그룹 순서대로 타야하기 때문에 검사를 해주자, 순서대로 탔는지 ! 

>여기서는 모든 정수를 검사할 필요 없이 [3,4,6,12] 4개의 항목만 검사하면 된다. 


* 여기서 시간 복잡도를 줄이기 위해 만약에 검사하다가     검사하는 값(ex [3]) 보다 그룹 별 태우는 수를 초과 할 경우, 뒤에는 볼 필요 없이 다음 값을 검사하도록 하여 시간 복잡도를 줄였다.   




[Incremental Project 문제]
===

In [113]:
import requests
import re


req = requests.get('https://www.koreatech.ac.kr/kor/Main.do')

source = req.text





source = re.sub('<script.*?>.*?</script>', '', source, 0, re.I|re.S)  #Script 사이 내용 삭제 

source =  re.sub(r'(?s)\<.*?\>[^\w\s]*', '', source).replace('\n', '')    # < > 사이 태그 문자 삭제 

source =  re.sub(r'(?s)\>.*?\>[^\w\s]*', '', source).replace('\n', '')     # > > 사이 문자 삭제 

source =  re.sub(r'(?s)\{.*?\}[^\w\s]*', '', source).replace('\n', '')     # {  }사이 문자 제거 

source =  re.sub('[-=+,#/\?:^$.@*\"※~&%ㆍ!』;}\\‘|\(\)\[\]\<\>`\'…》]', '', source)   #특수문자 제거 

#source = re.sub('[a-zA-z-0-9]', '', source)



source_list = source.split()    # 공백으로 문자들을 나누고 

#for i in source_list:
    #if (len(i)<20):       #띄어쓰기가 안된 엄청 긴 단어들은 표현 x 
        #print(i)          #주석처리 해제 하면 결과 출력 됩니다!  
        
   
        
print("모든 단어의 갯수는 : ",len(source_list), "개 입니다.")   

모든 단어의 갯수는 :  558 개 입니다.


<소감>
===

이번 과제를 해보면서 난이도가 있었던 것 같고 완벽하게 구현하지 못한 내용도 많다, 이번에 구현하지 못한 내용들은 파이썬 과목을 더욱 독학하고 복습해서 보완해보려고 노력해야겠다.